test/distributed/org/apache/cassandra/distributed/test/PutGetTest.java

```diff
@@ -0,0 +1,306 @@
+                {
+                    Object[] row = rows[i];
+                    validateRow(row, numRows, 0, c + i);
+                }
+            }
+        }
+    }
+
+    /**
+     * Simple put/get on multiple partitions with multiple rows, reading with IN queries.
+     */
+    @Test
+    public void testInQuery()
+    {
+        int numPartitions = 10;
+        int numRows = 10;
+
+        writeRows(numPartitions, numRows);
+
+        String query = "SELECT * FROM %s WHERE k IN (?, ?)";
+        for (int k = 0; k < numPartitions; k += 2)
+        {
+            Object[][] rows = readRows(query, k, k + 1);
+            Assert.assertEquals(numRows * 2, rows.length);
+
+            for (int i = 0; i < 2; i++)
+            {
+                for (int c = 0; c < numRows; c++)
+                {
+                    Object[] row = rows[i * numRows + c];
+                    validateRow(row, numRows, k + i, c);
+                }
+            }
+        }
+    }
+
+    /**
+     * Writes {@code numPartitions} with {@code numRows} each, with overrides in different sstables and memtables.
+     */
+    private void writeRows(int numPartitions, int numRows)
+    {
+        String update = format("UPDATE %s SET v=? WHERE k=? AND c=?");
+        ICoordinator coordinator = cluster.coordinator(1);
+        for (int k = 0; k < numPartitions; k++)
+        {
+            // insert all the partition rows in a single sstable
+            for (int c = 0; c < numRows; c++)
+                coordinator.execute(update, QUORUM, c, k, c);
+            cluster.forEach(i -> i.flush(KEYSPACE));
+
+            // override some rows in a second sstable
+            for (int c = 0; c < numRows; c += 2)
+                coordinator.execute(update, QUORUM, c + numRows, k, c);
+            cluster.forEach(i -> i.flush(KEYSPACE));
+
+            // override some rows only in memtable
+            for (int c = 0; c < numRows; c += 5)
+                coordinator.execute(update, QUORUM, c + numRows * 2, k, c);
+        }
+    }
+
+    /**
+     * Runs the specified query in all coordinators, with and without paging.
+     */
+    private Object[][] readRows(String query, Object... boundValues)
+    {
+        query = format(query);
+        Object[][] rows = null;
+
+        // verify that all coordinators return the same results for the query, regardless of paging
+        for (ICoordinator coordinator : cluster.coordinators())
+        {
+            for (boolean paging : BOOLEANS)
+            {
+                Object[][] readRows = paging
+                                      ? Iterators.toArray(coordinator.executeWithPaging(query, QUORUM, 1, boundValues),
+                                                          Object[].class)
+                                      : coordinator.execute(query, QUORUM, boundValues);
+                if (rows == null)
+                    rows = readRows;
+                else
+                    assertRows(rows, readRows);
```

Comment:
nit: Perhaps we should write an explicit message here indicating the assertion has failed because paged and non-paged versions of the result are different?

Reply:
Done. Indeed testing the different coordinators and paging here instead of just using Junit params makes things a bit more complex, but there is a significant performance gain over Junit parameterization.